### PR TITLE
tests: interrupt: add test cases of offloading job from isr

### PR DESCRIPTION
**Pull request by enjiamai**
Add 3 test cases to test different scenarios for offloading job from ISR, the test cases design as following: 

1. We spawn a thread to run then trigger an ISR context by irq_offload() or dynamic interrupt. 
2. In ISR handler, we offload the job to kwork, it could be several different or identical kworks.  
3. After the interrupt handler ending, the next running thread will depend on its priority,

The newly adding testcases are:
- test_isr_offload_job_multiple()
Validate the offloaded work executes immediately or not depends on its priority, and it offloads to different k_work. 

- test_isr_offload_job_identi()
Validate the offloaded work executes immediately or not depends on its priority, and it offloads to the identical k_work. 

- test_isr_offload_job()
Use dynamic interrupt instead of irq_offload() to verify the offloaded work. 

Signed-off-by: Enjia Mai <enjiax.mai@intel.com>

### CODE REVIEW

**Review by LeiW000 — APPROVED** [review]

File: tests/kernel/interrupt/src/interrupt_offload.c [hunk 1]
```diff
@@ -0,0 +1,266 @@
...
+ * This global variable control if the priority of offload job
+ * is greater than the original thread.
+ */
+static bool offload_job_prio_higher;
+
+static volatile int orig_t_keep_run;
+
+enum {
+	TEST_OFFLOAD_MULTI_JOBS,
+	TEST_OFFLOAD_INDETICAL_JOBS
```
Comment (maksimmasalski) [inline tests/kernel/interrupt/src/interrupt_offload.c:33]: Here is a misprint, IDENTICAL instead INDETICAL.

File: tests/kernel/interrupt/src/interrupt_offload.c [hunk 2]
```diff
@@ -0,0 +1,266 @@
...
+	k_sem_take(&start_sem, K_FOREVER);
+
+	for (int i = 0; i < NUM_WORK; i++) {
+
+		switch (case_type) {
+		case TEST_OFFLOAD_MULTI_JOBS:
+			trigger_offload_interrupt(real_irq,
+					(void *)&offload_work[i]);
+		break;
+		case TEST_OFFLOAD_INDETICAL_JOBS:
```
Comment (maksimmasalski) [inline tests/kernel/interrupt/src/interrupt_offload.c:173]: Here is the same misprint, IDENTICAL.

File: tests/kernel/interrupt/src/interrupt_offload.c [hunk 3]
```diff
@@ -0,0 +1,266 @@
...
+ *
+ * - If the priority of the original thread > offload job, offload jobs
+ *   will not execute immediately.
+ *
+ * We test this by irq_offload().
+ */
+void test_isr_offload_job_identi(void)
+{
+	offload_job_prio_higher = true;
+	run_test_offload(TEST_OFFLOAD_INDETICAL_JOBS, false);
```
Comment (maksimmasalski) [inline tests/kernel/interrupt/src/interrupt_offload.c:242]: Here too, IDENTICAL.

File: tests/kernel/interrupt/src/interrupt_offload.c [hunk 4]
```diff
@@ -0,0 +1,266 @@
...
+ */
+void test_isr_offload_job(void)
+{
+	if (!IS_ENABLED(CONFIG_DYNAMIC_INTERRUPTS)) {
+		ztest_test_skip();
+	}
+
+	offload_job_prio_higher = true;
+	run_test_offload(TEST_OFFLOAD_MULTI_JOBS, true);
+	run_test_offload(TEST_OFFLOAD_INDETICAL_JOBS, false);
```
Comment (maksimmasalski) [inline tests/kernel/interrupt/src/interrupt_offload.c:265]: IDENTICAL. Use search to find all misprints in that word.

**Review by maksimmasalski — CHANGES_REQUESTED** [review]

File: tests/kernel/interrupt/src/interrupt_offload.c (hunk 1 above)
Comment (enjiamai) [reply tests/kernel/interrupt/src/interrupt_offload.c]: Thanks for correcting this. Got it fixed.

File: tests/kernel/interrupt/src/interrupt_offload.c (hunk 2 above)
Comment (enjiamai) [reply tests/kernel/interrupt/src/interrupt_offload.c]: Thanks for correcting this. Got it fixed.

File: tests/kernel/interrupt/src/interrupt_offload.c (hunk 3 above)
Comment (enjiamai) [reply tests/kernel/interrupt/src/interrupt_offload.c]: Thanks for correcting this. Got it fixed.

File: tests/kernel/interrupt/src/interrupt_offload.c (hunk 4 above)
Comment (enjiamai) [reply tests/kernel/interrupt/src/interrupt_offload.c]: Thanks for correcting this. Got it fixed.

**Comment by enjiamai** [comment]
@maksimmasalski Thanks for your review, I have addressed the typo you mentioned.^^

**Comment by LeiW000** [comment]
@andyross, could you take a look at the PR?

File: tests/kernel/interrupt/src/interrupt_offload.c [hunk 5]
```diff
@@ -0,0 +1,266 @@
...
+
+struct interrupt_param {
+	struct k_work *work;
+};
+
+static struct interrupt_param irq_param;
+
+static void entry_offload_job(struct k_work *work)
+{
+	printk("doing offload job\n");
```
Comment (maksimmasalski) [inline tests/kernel/interrupt/src/interrupt_offload.c:44]: Please use TC_PRINT instead of printk in tests.

Comment (maksimmasalski) [reply tests/kernel/interrupt/src/interrupt_offload.c:44]: I think some prints mostly are for the developer purposes. Some of them you can just replace with C comments /* doing offload job */
I don't think we have to make dirty the output of the test.

Comment (enjiamai) [reply tests/kernel/interrupt/src/interrupt_offload.c]: Hi @maksimmasalski , I have addressed these two suggestions in my code, could you please take a look at it again, thanks!

File: tests/kernel/interrupt/src/interrupt_offload.c [hunk 6]
```diff
@@ -0,0 +1,265 @@
...
+
+	k_sem_give(&end_sem);
+}
+
+/* offload work to work queue */
+void isr_handler(const void *param)
+{
+	struct k_work *work = ((struct interrupt_param *)param)->work;
+
+	zassert_not_null(work, "NULL");
```
Comment (maksimmasalski) [inline tests/kernel/interrupt/src/interrupt_offload.c:63]: Could you give more details for that message? Message "NULL" is not descriptive.

Comment (enjiamai) [reply tests/kernel/interrupt/src/interrupt_offload.c]: Got it fixed, thanks!

**Comment by enjiamai** [comment]
Hi @maksimmasalski , thanks for point out the error in line 270 (the false shall be true), I have corrected it, thank you so much!

**Review by maksimmasalski — APPROVED** [review]

**Review by andyross — APPROVED** [review]

**Pull request merged**